crates/apollo-compiler/src/database/hir_db.rs

```diff
@@ -929,7 +942,12 @@ fn argument(argument: ast::Argument) -> Argument {
     }
 }
 
-fn value(val: ast::Value) -> Value {
+/// `expected_type` is the type of value that is expected to be returned,
+/// as far as known based on type system definitions.
+/// This affects values with integer syntax that should be interpreted as floats.
+///
+/// `None` means unknown, or type inference not fully implemented yet.
```

Comment:
This is sort of related to the conversation @goto-bus-stop and I were having. It feels like it's becoming a bit more important for us to be able to run extra compile-time checks, such as coercion or the type inference you're mentioning here. We are expecting the HIR to be this perfectly correct representation of the world, but in fact it can still be missing information or have incorrect types (like this argument value and its expected type). What it really means is that this hir_db is a bit overloaded.

I am wondering if in the long run we will have to create a further layer to perform some of these extra compile-time checks after some form of a HIR is created. This could either be done before or after the validation step.